librarian_background/rolling_deletion.py

```diff
@@ -0,0 +1,198 @@
+
+a) The file must have $N$ remote instances available throughout the network
+b) The checksums of those files must match the original checksum
+"""
+
+import time
+from datetime import datetime, timedelta, timezone
+
+from loguru import logger
+from schedule import CancelJob
+from sqlalchemy import select
+from sqlalchemy.orm import Session
+
+from librarian_server.api.validate import calculate_checksum_of_remote_copies
+from librarian_server.database import get_session
+from librarian_server.orm import Instance, Librarian, StoreMetadata
+
+from .task import Task
+
+
+class RollingDeletion(Task):
+    """
+    A background task that deletes _instances_ (not files!) that are older than
+    a certain age.
+    """
+
+    store_name: str
+    "Name of the store to delete instances from"
+    age_in_days: float
+    "Age of the instances to delete, in days; older files will be deleted if they pass the checks"
+
+    number_of_remote_copies: int = 3
+    "Number of remote copies that must be available to delete the file"
+    verify_downstream_checksums: bool = True
+    "Whether to verify the checksums of the remote copies"
+    mark_unavailable: bool = True
+    "Whether to mark the instances as unavailable after deletion, or to delete them (False)"
+    force_deletion: bool = True
+    "Whether to over-ride the deletion policy of instances"
+
+    def get_store(self, name: str, session: Session) -> StoreMetadata:
+        possible_metadata = session.query(StoreMetadata).filter_by(name=name).first()
+
+        if not possible_metadata:
+            raise ValueError(f"Store {name} does not exist.")
+
+        return possible_metadata
+
+    def on_call(self):
+        with get_session() as session:
+            return self.core(session=session)
+
+    def core(self, session: Session):
+        core_begin = datetime.now(timezone.utc)
+        age_cutoff = core_begin - timedelta(days=self.age_in_days)
+
+        try:
+            store = self.get_store(self.store_name, session)
+        except ValueError as e:
+            logger.error("Error getting store: {}, cannot continue; cancelling job", e)
+            return CancelJob
+
+        logger.info(
+            "Beginning rolling deletion for store {} (ID: {})", store.name, store.id
+        )
+
+        # Get the instances that are older than the age
+
+        logger.info(
+            "Querying for created_times later than {} UTC ({} local)",
+            age_cutoff,
+            age_cutoff.astimezone(),
+        )
+
+        query_begin = time.perf_counter()
+        stmt = select(Instance).filter(
+            Instance.store_id == store.id,
+            Instance.created_time < age_cutoff.astimezone(timezone.utc),
+            Instance.available == True,
+        )
+
+        instances = session.execute(stmt).scalars().all()
+        query_end = time.perf_counter()
+
+        logger.info("Queried for old instances in {} seconds", query_end - query_begin)
```

Comment:
Can we have this kind of information as debug? Also the timing information probes can be optional.

Reply:
I'd like to see the timing information in production where we usually don't run with all debugging statements on, hence why I left it as info. Is this generally bad practice?

Reply:
I generally prefer any messages that is showing information that is not about normal operations to be a debug message. In other words, info messages show the progress of the features of the system, while debug messages show the information that a developer is interested.

Reply:
I think given that the 'users' of this system are the people maintaining it, the database query performance is important enough on a day-to-day timescale that we want to keep it as info. I'll keep it as is.